### ISSUE TITLE
[NETWORK_SWITCH] Improve the reliability of the network_switch feature

### ISSUE THREAD

**Issue by benliao1** [pioneers/runtime#273]
The biggest technical holding back our competition from running smoothly is the inability for the robots to reliably switch networks between the team routers and `pioneers`. This has always been a problem that PiE has struggled with, but since the other software projects are much more stable now, this issue has become extremely glaring as during competitions everyone is just waiting on Runtime to figure out how to get the robots to connect.

In the past, we've tried setting up the `wpa_supplicant.conf` file to the team router network on a high priority and `pioneers` on a low priority, but we ran into issues of the robots not being able to connect their robots to the team routers at the season events and auto-connecting to `pioneers`, even when they were on the field. Under this plan, students who were up to run a match would come to inspection and show the inspectors that their team router was powered off (and therefore the high-priority team network would not be up). Then, when the inspectors power cycled the robot, the computer would, in theory, connect automatically connect to `pioneers`. 

We also tried, in the past, to get rid of the team routers at competition; i.e. have all the teams use `pioneers` both when working on their robots at the event and when running matches. This is also a good idea, but only if there is some easy way to be able to quickly change settings for the students at the end of the first day of final competition so that they can continue working on the robot after final competition / between the first and second day of final competition. Another issue we ran into was that since we needed to have `pioneers` running both at the field and the in the pits, we needed to have multiply access points (APs) running. This, however, would cause some robots to disconnect on the field during matches, because robots would roam between different APs during the match, depending on which AP the robot thought was providing a stronger signal. Perhaps this could be fixed if we disabled roaming on `pioneers`, but this could cause further issues would slow / spotty connections if the computer connects to the "wrong" AP on the field and then is not allowed to switch to a different AP with a stronger signal.

Last year, we had the electrical team implement a physical switch on the PDB called the network switch which would dictate which network the Raspberry Pi / Potato would connect to. The idea was that the since the switch would dictate which network the potato would connect to, it wouldn't be able to connect to the wrong network; if the computer wasn't able to connect to the desired network, it would just keep trying until it was able to connect. The idea is sound in theory, but we ran into issues at both spring and fall competition where the robot simply would not connect to pioneers even after multiple tries, often requiring Runtime members to manually ssh into the robots and do some combination of the following to get the robot to connect:

1. Reboot the robot
2. Reboot the networking interfaces
3. Connect to the team network again, then retry connecting to `pioneers`

This issue is to address this long-standing issue and try to come up with a viable, reliable solution for the networking problem. We desperately need to address this issue. Here are some ideas:

1. Resetting the network manager to a default state (for example, by forgetting all networks or something of that nature) every time we want to switch networks, so that the network manager has no memory of which network it connected to in the past. In my experience, forgetting networks on a computer is a good way to get a computer to connect more reliably to Wi-Fi in general; perhaps this technique could be tried here
2. Devising a system to use the team routers to run the competition (for example, perhaps we connect the team routers to the dinos and then have the dinos connect to shepherd via ethernet or some other fully wired network. This allows students to use the team routers for both working at home and running the game on the field
3. Devising a (working) system to only use `pioneers` at the season event. This would require us to address the roaming problem detailed in the description above.
4. Attaching an additional router on each robot on the field that will forward all traffic on the team router network to the `pioneers` network when a team is running a match on the field.
5. Improving on the network switch script to more intelligently attempt to detect faults in the network switching process and automatically take corrective action to give the script a higher chance of succeeding

This, the runtime status, the security, and the executor refactoring are the biggest items needing work in Runtime, but of those, I believe this is the most important since it is directly affecting our ability to run the competition smoothly. So this issue should be given the highest priority in the coming weeks / months and should be worked on immediately / tested extensively. Ideally, it should be complete at the end of winter break at the very latest so we will not need to do software updates on the student's hardware mid-season, like the disastrous Year 15 software updates.

**Comment by BrandonWong14** [comment]
The issue appears to have been solved by moving runtime back to Raspbian on Raspberry Pi from Ubuntu on Le Potatoes. The specific cause of the issue previously is unknown, but the error messaged made it appear to be an OS or packages on the OS issue when on Ubuntu.
